designs/credentialManagement/design.md

```diff
@@ -0,0 +1,115 @@
+# Credential Management
+
+## Abstract
+
+This document describes the process of managing and switching of AWS credentials within the toolkit.
+
+## Motivation
+
+AWS provides many different ways to retrieve the credentials required to make calls to AWS including but not limited to:
+credential/config files, environment variables, system properties, EC2 metadata APIs. You can also assume roles using 
+a source set of credentials or use MFA. Due to this large matrix of possibilities the toolkit strives to provide an 
+intuitive abstraction layer on top of some of these sources so that the implementation details are hidden and a 
+consistent user experience can be provided, while also enabling future expansion of new sources, including possibly by 
+third party plugins.
+
+## Specification
+
+This toolkit will introduce the following concepts:
```

Comment:
I'd like to see a diagram (UML, interaction, or something similar) to bring all these concepts together. The diagram could include one concrete class for each abstraction as well.